### ISSUE TITLE
[Bug]: Optimize status hangs in pending and minor

### ISSUE THREAD

**Issue by XBaith** [NetEase/arctic#955]
### What happened?

Hi, Optimize Iceberg has encountered two problems, and I haven't found the problem for a long time:
1. optimize group 1: The optimizer is still alive but the OptimizeQueue has no plan log, which means the queue no longer works? (It will no longer work after merging 2-3 tables successfully)

3. optimize group 2: optimize does not respond for a long time after the execution of minor, although its file size is very small. (Some tables succeed, some tables hang on or timeout retry)

And I'm not sure the bug ocurred by my change which I drop all status in `BaseFileScan` to reduce jvm heap memory(BTW, It does save a lot of memory. If this has no impact on optimization, I will trigger a new PR for this.). 

### Affects Versions

master/0.4.x

### What engines are you seeing the problem on?

AMS

### How to reproduce

just optimize a certain number of iceberg tables

### Anything else

![image](https://user-images.githubusercontent.com/54210725/208618448-5f34d7fb-a34b-4604-a6b9-2788d3bf619e.png)
![image](https://user-images.githubusercontent.com/54210725/208618547-90442eab-9110-4df8-b6c4-088e3cb2097a.png)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by majin1102** [comment]
cc @wangtaohz @hzluting

**Comment by hzluting** [comment]
Hello, thanks for your feedback.
For the first problem, if you can confirm the optimizer Flink task is alive, and the plan log of this optimize queue doesn't output(maybe there are some error message in ams-error.log). You can use jstack for the AMS and search a thread named "Optimize Plan Thread Queue-xx(xx is the optimize queue id)",  maybe this thread blocked by some operations cost much time, such as the hive sync in table refresh.
For the second problem, the optimize status of the table isn't accurate. Once the table has optimize task and the optimize status is optimizing(minor or major). And we will update the logic of optimize status :
1. optimizing(minor or major): Only the table has the optimize task in executing
2. pending: The table has no executing optimize task but has pending or init optimize task, or the table hasn't optimize task now but needs to optimize
3. idle: The table has no executing and pending optimize task

**Comment by majin1102** [comment]
> Hello, thanks for your feedback. For the first problem, if you can confirm the optimizer Flink task is alive, and the plan log of this optimize queue doesn't output(maybe there are some error message in ams-error.log). You can use jstack for the AMS and search a thread named "Optimize Plan Thread Queue-xx(xx is the optimize queue id)", maybe this thread blocked by some operations cost much time, such as the hive sync in table refresh. For the second problem, the optimize status of the table isn't accurate. Once the table has optimize task and the optimize status is optimizing(minor or major). And we will update the logic of optimize status :
> 
> 1. optimizing(minor or major): Only the table has the optimize task in executing
> 2. pending: The table has no executing optimize task but has pending or init optimize task, or the table hasn't optimize task now but needs to optimize
> 3. idle: The table has no executing and pending optimize task

The second issue, I don't get it, is this a bug or designing defects?

**Comment by hzluting** [comment]
> > Hello, thanks for your feedback. For the first problem, if you can confirm the optimizer Flink task is alive, and the plan log of this optimize queue doesn't output(maybe there are some error message in ams-error.log). You can use jstack for the AMS and search a thread named "Optimize Plan Thread Queue-xx(xx is the optimize queue id)", maybe this thread blocked by some operations cost much time, such as the hive sync in table refresh. For the second problem, the optimize status of the table isn't accurate. Once the table has optimize task and the optimize status is optimizing(minor or major). And we will update the logic of optimize status :
> > 
> > 1. optimizing(minor or major): Only the table has the optimize task in executing
> > 2. pending: The table has no executing optimize task but has pending or init optimize task, or the table hasn't optimize task now but needs to optimize
> > 3. idle: The table has no executing and pending optimize task
> 
> The second issue, I don't get it, is this a bug or designing defects?

This is designing defects now. After, we will optimize it with the logic that I said above.

**Comment by majin1102** [comment]
> > > Hello, thanks for your feedback. For the first problem, if you can confirm the optimizer Flink task is alive, and the plan log of this optimize queue doesn't output(maybe there are some error message in ams-error.log). You can use jstack for the AMS and search a thread named "Optimize Plan Thread Queue-xx(xx is the optimize queue id)", maybe this thread blocked by some operations cost much time, such as the hive sync in table refresh. For the second problem, the optimize status of the table isn't accurate. Once the table has optimize task and the optimize status is optimizing(minor or major). And we will update the logic of optimize status :
> > > 
> > > 1. optimizing(minor or major): Only the table has the optimize task in executing
> > > 2. pending: The table has no executing optimize task but has pending or init optimize task, or the table hasn't optimize task now but needs to optimize
> > > 3. idle: The table has no executing and pending optimize task
> > 
> > 
> > The second issue, I don't get it, is this a bug or designing defects?
> 
> This is designing defects now. After, we will optimize it with the logic that I said above.

OK， we shall make an improvement thread and I will follow this up. Thanks for @XBaith feeding back

**Comment by majin1102** [comment]
> > > > Hello, thanks for your feedback. For the first problem, if you can confirm the optimizer Flink task is alive, and the plan log of this optimize queue doesn't output(maybe there are some error message in ams-error.log). You can use jstack for the AMS and search a thread named "Optimize Plan Thread Queue-xx(xx is the optimize queue id)", maybe this thread blocked by some operations cost much time, such as the hive sync in table refresh. For the second problem, the optimize status of the table isn't accurate. Once the table has optimize task and the optimize status is optimizing(minor or major). And we will update the logic of optimize status :
> > > > 
> > > > 1. optimizing(minor or major): Only the table has the optimize task in executing
> > > > 2. pending: The table has no executing optimize task but has pending or init optimize task, or the table hasn't optimize task now but needs to optimize
> > > > 3. idle: The table has no executing and pending optimize task
> > > 
> > > 
> > > The second issue, I don't get it, is this a bug or designing defects?
> > 
> > 
> > This is designing defects now. After, we will optimize it with the logic that I said above.
> 
> OK， we shall make an improvement thread #959  and I will follow this up. Thanks for @XBaith feeding back

**Comment by XBaith** [comment]
> > 
> 
> @XBaith Hello, did you run jstack for AMS many times, and all print the same status of the thread named Optimize Plan Thread Queue-2? If only run jstack one time, it maybe can't confirm this thread hangs or is blocked. So, please confirm it first. Thanks.

Yes. It still locked

**Comment by wangtaohz** [comment]
> Looks like this thread was locked for some reason.

According to the result of jstack, the thread `Optimize Plan Thread Queue-2` was not locked, but still RUNNABLE. 
The ` - locked <0x00000006c82964e8> (a java.lang.Throwable)` means it locked the object of Throwable with `synchronized` as follows
```
// java.lang.Throwable
    private synchronized StackTraceElement[] getOurStackTrace() {
        // Initialize stack trace field with information from
        // backtrace if this is the first call to this method
        if (stackTrace == UNASSIGNED_STACK ||
            (stackTrace == null && backtrace != null) /* Out of protocol state */) {
            int depth = getStackTraceDepth();
            stackTrace = new StackTraceElement[depth];
            for (int i=0; i < depth; i++)
                stackTrace[i] = getStackTraceElement(i);
        } else if (stackTrace == null) {
            return UNASSIGNED_STACK;
        }
        return stackTrace;
    }
```

and this method was called only for the reason of printing debug log.

```
//org.apache.hadoop.security.UserGroupInformation

  private void logPrivilegedAction(Subject subject, Object action) {
    if (LOG.isDebugEnabled()) {
      // would be nice if action included a descriptive toString()
      String where = new Throwable().getStackTrace()[2].toString();
      LOG.debug("PrivilegedAction as:"+this+" from:"+where);
    }
  }
```


Do you mean that you run jstack many times, and they all printed the same result?

If so, we gonna go deep dive into why the native method `java.lang.Throwable.getStackTraceElement` was blocked.

**Comment by XBaith** [comment]
> > Looks like this thread was locked for some reason.
> 
> According to the result of jstack, the thread `Optimize Plan Thread Queue-2` was not locked, but still RUNNABLE. The ` - locked <0x00000006c82964e8> (a java.lang.Throwable)` means it locked the object of Throwable with `synchronized` as follows
> 
> ```
> // java.lang.Throwable
>     private synchronized StackTraceElement[] getOurStackTrace() {
>         // Initialize stack trace field with information from
>         // backtrace if this is the first call to this method
>         if (stackTrace == UNASSIGNED_STACK ||
>             (stackTrace == null && backtrace != null) /* Out of protocol state */) {
>             int depth = getStackTraceDepth();
>             stackTrace = new StackTraceElement[depth];
>             for (int i=0; i < depth; i++)
>                 stackTrace[i] = getStackTraceElement(i);
>         } else if (stackTrace == null) {
>             return UNASSIGNED_STACK;
>         }
>         return stackTrace;
>     }
> ```
> 
> and this method was called only for the reason of printing debug log.
> 
> ```
> //org.apache.hadoop.security.UserGroupInformation
> 
>   private void logPrivilegedAction(Subject subject, Object action) {
>     if (LOG.isDebugEnabled()) {
>       // would be nice if action included a descriptive toString()
>       String where = new Throwable().getStackTrace()[2].toString();
>       LOG.debug("PrivilegedAction as:"+this+" from:"+where);
>     }
>   }
> ```
> 
> Do you mean that you run jstack many times, and they all printed the same result?
> 
> If so, we gonna go deep dive into why the native method `java.lang.Throwable.getStackTraceElement` was blocked.

Good answer. I do reset LOG level for avoiding print too many icebert logs when check in `OptimizeQueueService`. It's the issue related this reason?
```
org.apache.log4j.Logger.getLogger(org.apache.iceberg.BaseMetastoreCatalog.class).setLevel(Level.WARN);
    org.apache.log4j.Logger.getLogger(org.apache.iceberg.BaseMetastoreTableOperations.class).setLevel(Level.WARN);
    org.apache.log4j.Logger.getLogger(org.apache.iceberg.DataTableScan.class).setLevel(Level.WARN);
```

**Comment by wangtaohz** [comment]
> Good answer. I do reset LOG level for avoiding print too many icebert logs when check in OptimizeQueueService. It's the issue related this reason?

It may not. I can hardly find the relations between them.

**Comment by hzluting** [comment]
@XBaith @wangtaohz I think this method needs to optimize. Every time call the table.spec() will call ArcticFileIo.doAs() and the getOurStackTrace().

**Comment by wangtaohz** [comment]
> @XBaith @wangtaohz I think this method needs to optimize. Each DataFile build will call this method and every time call the table.spec() will call ArcticFileIo.doAs() and the getOurStackTrace().

Yes, I fixed it with #962. Please take a look.

**Comment by wangtaohz** [comment]
> I ran jstack a few times, and I found that the Plan thread is in the "sync" state most of the time, when the state switches to running and then switches back to "sync" state very quickly.

Thanks for your reply, but I'm not quite sure what the `"sync" state` means exactly.
Maybe the raw results of jstack will be much more helpful.
 And I have fixed the problem that invokes `getOurStackTrace()` too many times, with #962. It will help if this is the bottleneck.

**Comment by zhoujinsong** [comment]
@XBaith  Does #963 Fix this problem?

**Comment by XBaith** [comment]
> @XBaith Did #963 fix this problem?

@zhoujinsong Fix hang in pending issue but not minor status

**Comment by zhoujinsong** [comment]
> > @XBaith Did #963 fix this problem?
> 
> @zhoujinsong Fix hang in pending issue but not minor status

Okay, another PR #959 is trying to solve the minor status problem.

**Comment by wangtaohz** [comment]
> In one case, if the optimization task fails and is retried more than 5 times, the state of table optimization will be suspended. If we want to re-optimize it, we have to clean up the tasks, files and tables in the database then restart AMS.

In this case, the only way is to clean up these tasks, however, task clean-up is very cumbersome now. We are going to clean up these tasks automatically and leave an `optimize-history` with some error messages, then the `Optimizing` can keep going.

The AMS is under refactoring, after that the auto-clean-up feature will be included.
